lib/docker/api.ex

```diff
@@ -22,16 +22,48 @@ defmodule Testcontainers.Docker.Api do
     end
   end
 
-  def pull_image(image, conn) when is_binary(image) do
-    case Api.Image.image_create(conn, fromImage: image) do
-      {:ok, %Tesla.Env{status: 200}} ->
-        {:ok, nil}
+  def list_images(conn) do
+    case DockerEngineAPI.Api.Image.image_list(conn) do
+      {:ok, %DockerEngineAPI.Model.ErrorResponse{} = error} ->
+        {:error, {:failed_to_list_images, error}}
+
+      {:ok, list} ->
+        list
+        # |> Enum.map(fn x -> Map.get(x, :RepoTags) end)
+        |> Enum.flat_map(fn x -> Map.get(x, :RepoTags, []) || [] end)
+        |> Enum.uniq()
+        |> Enum.sort()
+    end
+  end
 
-      {:error, %Tesla.Env{status: other}} ->
-        {:error, {:http_error, other}}
+  def ping(conn) do
+    {:ok, %{body: body}} = DockerEngineAPI.Api.System.system_ping(conn)
+    body
+  end
 
-      {:ok, %DockerEngineAPI.Model.ErrorResponse{} = error} ->
-        {:error, {:failed_to_pull_image, error}}
+  def version(conn) do
+    {:ok, info} = DockerEngineAPI.Api.System.system_version(conn)
+    info
+  end
+
+  def pull_image(config, conn) when is_binary(config.image) do
+    image = config.image
+    auth = Map.get(config, :auth, nil)
+    is_local_image = Map.get(config, :is_local_image, false)
```

Comment:
If these are optional values, its more idiomatic to make this a Keyword list and use `Keyword.validate!`, with defaults e.g.

Also because you always expect an image, keep that as its own argument

```elixir
def pull_image(image, conn, opts \\ []) do
  opts = Keyword.validate!(opts, [:auth, local_image?: false])
```

Note that I did two changes here.

1. Changed the name to `local_image?` which is more idiomatic
2. Used `Keyword.validate!` which raises a nice error message if you pass an invalid option, and provides a default for `local_image?`

Reply:
@davydog187 That's what I had initially, but take a look at how `pull_image` is used. It's part of the builder pipeline and most other functions take the entire configuration. I mainly implemented it like this for consistency with the other functions in the pipeline.